File: experiments/android-readplace-poc/app/src/main/kotlin/com/readplace/poc/app/ReadingListViewModel.kt
package com.readplace.poc.app

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.readplace.poc.core.ApiError
import com.readplace.poc.core.ApiException
import com.readplace.poc.core.Article
import com.readplace.poc.core.QueuePage
import com.readplace.poc.core.ReadplaceApi
import com.readplace.poc.core.SirenAction
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Holds reading-list state for the Compose UI — the analogue of the iOS POC's
 * `ReadingListViewModel`. Network calls run on [Dispatchers.IO]; the synchronous core
 * stays off the main thread. State is exposed as Compose snapshot state so screens
 * recompose on change.
 */
class ReadingListViewModel(
	private val api: ReadplaceApi,
	private val onSessionExpired: () -> Unit,
) {
	var articles by mutableStateOf<List<Article>>(emptyList())
		private set
	var isLoading by mutableStateOf(false)
		private set
	var isSaving by mutableStateOf(false)
		private set
	var errorText by mutableStateOf<String?>(null)
	var warningText by mutableStateOf<String?>(null)

	private var nextHref: String? = null
	private var saveArticleAction: SirenAction? = null

	val hasMore: Boolean get() = nextHref != null

	suspend fun loadIfNeeded() {
		if (articles.isEmpty()) fetchFirstPage()
	}

	suspend fun refresh() = fetchFirstPage()

	private suspend fun fetchFirstPage() {
		isLoading = true
		errorText = null
		runCatching { withContext(Dispatchers.IO) { api.loadQueue() } }
			.onSuccess { applyPage(it, replacing = true) }
			.onFailure { handle(it) }
		isLoading = false
	}

	suspend fun loadMore() {
		val next = nextHref ?: return
		runCatching { withContext(Dispatchers.IO) { api.loadQueue(next) } }
			.onSuccess { applyPage(it, replacing = false) }
			.onFailure { handle(it) }
	}

	suspend fun saveUrl(rawUrl: String) {
		val trimmed = rawUrl.trim()
		val action = saveArticleAction
		if (trimmed.isEmpty() || action == null) return
		isSaving = true
		errorText = null
		runCatching { withContext(Dispatchers.IO) { api.saveArticle(action, trimmed) } }
			.onSuccess { fetchFirstPage() }
			.onFailure { handle(it) }
		isSaving = false
	}

	suspend fun delete(article: Article) {
		val href = article.deleteHref ?: return
		val snapshot = articles
		articles = articles.filterNot { it.id == article.id }
		runCatching { withContext(Dispatchers.IO) { api.delete(href) } }
			.onSuccess { applyPage(it, replacing = true) }
			.onFailure { error ->
				val alreadyGone = error is ApiException && error.error is ApiError.NotFound
				if (!alreadyGone) {
					articles = snapshot
					handle(error)
				}
			}
	}

	private fun applyPage(page: QueuePage, replacing: Boolean) {
		articles = if (replacing) {
			page.articles
		} else {
			val existing = articles.mapTo(HashSet()) { it.id }
			articles + page.articles.filter { it.id !in existing }
		}
		nextHref = page.nextHref
		page.saveArticleAction?.let { saveArticleAction = it }
		warningText = page.warning?.message
	}

	private fun handle(error: Throwable) {
		val apiError = (error as? ApiException)?.error
		if (apiError is ApiError.Unauthorized || apiError is ApiError.NoToken) {
			onSessionExpired()
		} else {
			errorText = apiError?.message ?: error.message ?: "Something went wrong."
		}
	}
}

File: projects/ios-readplace-poc/App/ReadingListViewModel.swift
import Foundation

/// A locked-account refusal surfaced to the UI: the server's message, which
/// itself names the address to email. The refusal models no action.
struct AccountLockout: Equatable {
	let message: String
}

@MainActor
final class ReadingListViewModel: ObservableObject {
	@Published private(set) var articles: [Article] = []
	@Published private(set) var isLoading = false
	@Published private(set) var isSaving = false
	@Published private(set) var hasMore = false
	@Published var errorText: String?
	@Published var warningText: String?
	@Published var lockout: AccountLockout?

	private var nextHref: String?
	private var isLoadingMore = false
	private var saveArticleAction: SirenAction?

	private let api: ReadplaceAPI
	private let onSessionExpired: () -> Void

	init(api: ReadplaceAPI, onSessionExpired: @escaping () -> Void) {
		self.api = api
		self.onSessionExpired = onSessionExpired
	}

	func loadIfNeeded() async {
		guard articles.isEmpty else { return }
		await fetchFirstPage()
	}

	func refresh() async {
		await fetchFirstPage()
	}

	private func fetchFirstPage() async {
		isLoading = true
		errorText = nil
		do {
			let page = try await api.loadQueue()
			apply(page, replacing: true)
		} catch {
			handle(error)
		}
		isLoading = false
	}

	func loadMore() async {
		guard let next = nextHref, !isLoadingMore else { return }
		isLoadingMore = true
		do {
			let page = try await api.loadQueue(path: next)
			apply(page, replacing: false)
		} catch {
			handle(error)
		}
		isLoadingMore = false
	}

	func delete(_ article: Article) async {
		guard let href = article.deleteHref else { return }
		let snapshot = articles
		articles.removeAll { $0.id == article.id }
		do {
			let page = try await api.delete(href: href)
			apply(page, replacing: true)
		} catch APIError.notFound {
			// Already gone server-side; keep it removed.
		} catch {
			articles = snapshot
			handle(error)
		}
	}

	func saveURL(_ rawURL: String) async {
		let trimmed = rawURL.trimmingCharacters(in: .whitespacesAndNewlines)
		guard !trimmed.isEmpty, let action = saveArticleAction else { return }
		isSaving = true
		errorText = nil
		lockout = nil
		do {
			_ = try await api.saveArticle(action: action, url: trimmed)
			await fetchFirstPage()
		} catch {
			handle(error)
		}
		isSaving = false
	}

	private func apply(_ page: QueuePage, replacing: Bool) {
		if replacing {
			articles = page.articles
		} else {
			let existing = Set(articles.map(\.id))
			articles += page.articles.filter { !existing.contains($0.id) }
		}
		nextHref = page.nextHref
		hasMore = page.nextHref != nil
		if let save = page.saveArticleAction { saveArticleAction = save }
		warningText = page.warning?.message
	}

	private func handle(_ error: Error) {
		switch error {
		case APIError.unauthorized, APIError.noToken:
			onSessionExpired()
		case let APIError.accountLocked(message):
			lockout = AccountLockout(message: message)
		default:
			errorText = (error as? LocalizedError)?.errorDescription ?? error.localizedDescription
		}
	}
}
